examples/cpp_api/png_ingestion_webp.cc

```diff
@@ -0,0 +1,369 @@
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @section DESCRIPTION
+ *
+ * This is a simple ingestor program for TileDB that ingests PNG data into an
+ * array using WebP filter
+ */
+
+#include <cassert>
+#include <cstdio>
+
+#include <png.h>
+// Note: on some macOS platforms with a brew-installed libpng, use this instead:
+//#include <libpng16/png.h>
+
+// Include the TileDB C++ API headers
+#include <tiledb/tiledb>
+
+using namespace tiledb;
+
+std::vector<uint8_t> read, write;
+
+// Colorspace stride
+const unsigned colorspace = TILEDB_WEBP_RGBA;
+const unsigned pixel_depth = colorspace < TILEDB_WEBP_RGBA ? 3 : 4;
+const float quality_factor = 100.0f;
+const bool lossless = true;
+
+/**
+ * Reads a .png file at the given path and returns a vector of pointers to
+ * the pixel data in each row. The caller must free the row pointers.
+ *
+ * This is a modified version of: https://gist.github.com/niw/5963798
+ * "How to read and write PNG file using libpng"
+ * (C) 2002-2010 Guillaume Cottenceau
+ * Redistributed under the X11 license.
+ */
+std::vector<uint8_t*> read_png(
+    const std::string& path, unsigned* width, unsigned* height) {
+  std::vector<uint8_t*> row_pointers;
+
+  // Get the image info.
+  auto fp = fopen(path.c_str(), "rb");
+  png_structp png =
+      png_create_read_struct(PNG_LIBPNG_VER_STRING, NULL, NULL, NULL);
+  png_infop info = png_create_info_struct(png);
+  setjmp(png_jmpbuf(png));
+  png_init_io(png, fp);
+  png_read_info(png, info);
+
+  *width = png_get_image_width(png, info);
+  *height = png_get_image_height(png, info);
+  uint8_t color_type = png_get_color_type(png, info),
+          bit_depth = png_get_bit_depth(png, info);
+
+  // Read any color_type into 8bit depth, RGBA format.
+  // See http://www.libpng.org/pub/png/libpng-manual.txt
+  if (bit_depth == 16)
+    png_set_strip_16(png);
+
+  if (color_type == PNG_COLOR_TYPE_PALETTE)
+    png_set_palette_to_rgb(png);
+
+  // PNG_COLOR_TYPE_GRAY_ALPHA is always 8 or 16bit depth.
+  if (color_type == PNG_COLOR_TYPE_GRAY && bit_depth < 8)
+    png_set_expand_gray_1_2_4_to_8(png);
+
+  if (png_get_valid(png, info, PNG_INFO_tRNS))
+    png_set_tRNS_to_alpha(png);
+
+  // If example is ran with RGBA ensure alpha is present
```

Comment:
Choose whether all your comments end with a period or none of them--usually I'd prefer the former, since it seems more formal.